src/Datadog.Trace/Agent/MovingAverageKeepRateCalculator.cs

```diff
@@ -0,0 +1,165 @@
+            {
+                throw new ArgumentOutOfRangeException(nameof(count));
+            }
+
+            Interlocked.Add(ref _latestKeeps, count);
+        }
+
+        /// <summary>
+        /// Increment the number of dropped traces
+        /// </summary>
+        public void IncrementDrops(int count)
+        {
+            if (count < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(count));
+            }
+
+            Interlocked.Add(ref _latestDrops, count);
+        }
+
+        /// <summary>
+        /// Get the current keep rate for traces
+        /// </summary>
+        public double GetKeepRate()
+        {
+            var totals = Interlocked.Read(ref _totals);
+            UnpackBits(totals, out var totalDropped, out var totalCreated);
+            if (totalCreated == 0)
+            {
+                return 0;
+            }
+
+            return 1 - ((double)totalDropped / totalCreated);
+        }
+
+        /// <summary>
+        /// Stop updating the buckets. The current Keep rate can continue to be read.
+        /// </summary>
+        public void CancelUpdates()
+        {
+            _processExit.TrySetResult(true);
+        }
+
+        /// <summary>
+        /// Update the current rate. Internal for testing only. Should not be called in normal usage.
+        /// </summary>
+        internal void UpdateBucket()
+        {
+            var index = _index;
+            var previousDropped = _dropped[_index];
+            var previousCreated = _created[_index];
+
+            var latestDropped = Math.Min(Interlocked.Exchange(ref _latestDrops, 0), uint.MaxValue);
+            var latestCreated = Math.Min(Interlocked.Exchange(ref _latestKeeps, 0) + latestDropped, uint.MaxValue);
+
+            UnpackBits(_totals, out var previousTotalDropped, out var previousTotalCreated);
+
+            var newTotalDropped = (uint)Math.Min(((long)previousTotalDropped) - previousDropped + latestDropped, uint.MaxValue);
+            var newTotalCreated = (uint)Math.Min(((long)previousTotalCreated) - previousCreated + latestCreated, uint.MaxValue);
+
+            // Pack the totals into a single value we can read and write atomically
+            var packedTotal = PackBits(newTotalDropped, newTotalCreated);
+            Interlocked.Exchange(ref _totals, packedTotal);
+
+            _dropped[_index] = (uint)latestDropped;
+            _created[_index] = (uint)latestCreated;
+            _index = (index + 1) % _windowSize;
+        }
+
+        private async Task UpdateBucketTaskLoopAsync()
+        {
+            while (true)
+            {
+                if (_processExit.Task.IsCompleted)
+                {
+                    return;
+                }
+
+                UpdateBucket();
+
+                await Task.WhenAny(
+                               Task.Delay(_bucketDuration),
+                               _processExit.Task)
+                          .ConfigureAwait(false);
```

Comment:
I see this a lot in the repo too, In this case for old frameworks `Task.WhenAny` uses a `params` so create a new array on the call, and also it does a defensive copy internally, so 2 arrays per call, and because we are in a loop we allocate that in each pass, we could create a Task[2] array before the while and reuse for each loop. so we reduce to 1 array per call.

This problem doesn't happen in newer runtime versions. 

It just an allocation comment, feel free to ignore it if you want.

Reply:
Good to know, thanks! 

Given we know there's a single thread here, can we pre-allocate the array as a field, avoid the allocation entirely?

```csharp
private readonly Task[] _updateLoopTasks = new Task[2];

 private async Task UpdateBucketTaskLoopAsync()
{
    _updateLoopTasks[0] = _processExit.Task;

    while (true)
    {
        if (_processExit.Task.IsCompleted)
        {
            return;
        }

        UpdateBucket();

       _updateLoopTasks[1] =  Task.Delay(_bucketDuration)

        await Task.WhenAny(_updateLoopTasks).ConfigureAwait(false);
    }
}
```

Reply:
I think we can but that will add an allocation to newer framework versions 😄 (in new versions there is an overload for 2 Task parameters [here](https://source.dot.net/#System.Private.CoreLib/Task.cs,6054))

And the target for the optimization is this one: https://referencesource.microsoft.com/#mscorlib/system/threading/Tasks/Task.cs,6469

Maybe we can leave this kind of optimization in another PR...

Reply:
Oh wait, I didn't read your comment properly, I suggested an uglier version of what you suggested 🤦‍♂️.  I dunno, I'm inclined to fix it now as it won't get fixed otherwise 😉 Also, AFAICT, the [newer 2 Task Parameter version is only available in .NET 5](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.whenany?view=net-5.0#System_Threading_Tasks_Task_WhenAny_System_Threading_Tasks_Task_System_Threading_Tasks_Task_), so most versions could do with the optimisation?

Reply:
+ we don't explicitly compile for .NET 5.0 (maybe we should?), so we'll not get that optimisation currently anyway?

Reply:
oh you're right!... ha ha I see you already implement it.

Reply:
although we must be careful with `NET5_0_OR_GREATER` and the CI sdk version :D

Reply:
> although we must be careful with NET5_0_OR_GREATER and the CI sdk version :D

Yep, but as we're not compiling for .NET 5 at all at the moment, I feel like we can cross that bridge when we come to it 😉